File: c/linear_congruential_gen/gen.c

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

int main() {
    const int32_t SEED = 0;
    int32_t state = SEED;
    int32_t val = 1;
    int64_t period = 0;

    while (val != SEED){
        val = ((state * 1103515245) + 12345) & 0x7fffffff;
        if (val < 0x100){
            printf("period %"PRId64" state %x, val %x\n", period, state, val);
        }
        period++;
        state = val;
    }

    printf("%"PRId64"\n", period);

}

File: go/linear_congruential_gen/gen.go
package main

import "fmt"

func main() {
    
    const SEED int = 0
    var state int = SEED
    var val int = 1
    var period int = 0

    for val != SEED {
        val = ((state * 1103515245) + 12345) & 0x7fffffff
        if val < 0x100 {
            fmt.Printf("period %d state %x, val %x\n", period, state, val)
        }
        period++
        state = val
    }

    fmt.Printf("%d\n", period)

}
